inference-engine/thirdparty/clDNN/src/gpu/loop_gpu.cpp

```diff
@@ -0,0 +1,478 @@
+/*
+// Copyright (c) 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "loop_inst.h"
+#include "network_impl.h"
+#include "implementation_map.h"
+#include "math_utils.h"
+#include "register_gpu.hpp"
+#include "mutable_data_inst.h"
+#include "input_layout_inst.h"
+#include "memory_impl.h"
+#include <vector>
+#include <algorithm>
+
+namespace cldnn {
+namespace gpu {
+struct loop_gpu : typed_primitive_impl<loop> {
+    const loop_node& node;
+
+    explicit loop_gpu(const loop_node& node) : node(node) {}
+
+    static memory_impl::ptr get_outer_output_memory(const loop_inst& instance, const primitive_id& external_id) {
+        if (external_id == instance.id()) {
+            // main output
+            memory_impl& memory = instance.output_memory();
+            return (memory_impl::ptr) &memory;
+        } else {
+            // other primitives
+            const auto outputPrim = instance.get_network().get_primitive(external_id);
+            memory_impl& memory = outputPrim->output_memory();
+            return (memory_impl::ptr) &memory;
+        }
+    }
+
+    std::vector<loop::primitive_mapping> find_primitive_mappings(const primitive_id& internal_id) {
+        const auto& input_mappings = node.get_input_mappings();
+        const auto& output_mappings = node.get_output_mappings();
+
+        std::vector<loop::primitive_mapping> ret;
+        for (const auto& pm : input_mappings) {
+            if (pm.internal_id == internal_id) {
+                ret.push_back(pm);
+            }
+        }
+        for (const auto& pm : output_mappings) {
+            if (pm.internal_id == internal_id) {
+                ret.push_back(pm);
+            }
+        }
```

Comment:
I had a question for this.. can internal_id be mapped to multiple external_id ? 
(just out of curiosity whether you obeserved such kind of network or not)

Reply:
I only observed the models which has LSTMCell + TI, so I am not sure there is the case that internal_id is mapped to multiple externnal_id.

Reply:
even though there are, how about changing the map to unordered_map ?

Reply:
(not hightest priority)